src/ray/pubsub/subscriber.cc

```diff
@@ -18,174 +18,272 @@ namespace ray {
+bool SubscriberChannel<MessageID>::SubscriptionExists(const PublisherID &publisher_id) {
+  return subscription_map_.count(publisher_id);
+}
+
+///////////////////////////////////////////////////////////////////////////////
+/// Subscriber
+///////////////////////////////////////////////////////////////////////////////
+
+inline std::shared_ptr<SubscribeChannelInterface> Subscriber::Channel(
+    const rpc::ChannelType channel_type) const {
+  const auto it = channels_.find(channel_type);
+  RAY_CHECK(it != channels_.end()) << "Unknown channel: " << channel_type;
+  return it->second;
+}
+
+void Subscriber::Subscribe(const rpc::ChannelType channel_type,
+                           const rpc::Address &publisher_address,
+                           const std::string &message_id_binary,
+                           SubscriptionCallback subscription_callback,
+                           SubscriptionFailureCallback subscription_failure_callback) {
+  Channel(channel_type)
+      ->Subscribe(publisher_address, message_id_binary, std::move(subscription_callback),
+                  std::move(subscription_failure_callback));
+
+  // Make a long polling connection if we never made the one with this publisher for
+  // pubsub operations.
+  const auto publisher_id = PublisherID::FromBinary(publisher_address.worker_id());
+  auto publishers_connected_it = publishers_connected_.find(publisher_id);
+  if (publishers_connected_it == publishers_connected_.end()) {
+    publishers_connected_.emplace(publisher_id);
+    rpc::Address subscriber_address;
+    subscriber_address.set_raylet_id(subscriber_id_.Binary());
+    subscriber_address.set_ip_address(subscriber_address_);
+    subscriber_address.set_port(subscriber_port_);
+    MakeLongPollingPubsubConnection(publisher_address, subscriber_address);
+  }
+}
+
+bool Subscriber::Unsubscribe(const rpc::ChannelType channel_type,
+                             const rpc::Address &publisher_address,
+                             const std::string &message_id_binary) {
+  return Channel(channel_type)->Unsubscribe(publisher_address, message_id_binary);
+}
+
+void Subscriber::MakeLongPollingPubsubConnection(const rpc::Address &publisher_address,
+                                                 const rpc::Address &subscriber_address) {
+  const auto publisher_id = PublisherID::FromBinary(publisher_address.worker_id());
+  RAY_LOG(DEBUG) << "Make a long polling request to " << publisher_id;
+  auto publisher_client = publisher_client_pool_.GetOrConnect(publisher_address);
+  rpc::PubsubLongPollingRequest long_polling_request;
+  long_polling_request.mutable_subscriber_address()->CopyFrom(subscriber_address);
+
+  publisher_client->PubsubLongPolling(
+      long_polling_request, [this, publisher_address, subscriber_address](
+                                Status status, const rpc::PubsubLongPollingReply &reply) {
+        HandleLongPollingResponse(publisher_address, subscriber_address, status, reply);
+      });
```

Comment:
Nit: Could `std::bind` this to reduce pass-through boilerplate:
```suggestion
      long_polling_request, std::bind(&HandleLongPollingResponse, this, publisher_address, subscriber_address, std::placeholders::_1, std::placeholders::_2));
```

Reply:
Personally thinking this placeholder stuff is hard to read, so I will just keep the current style.